source/AI.cpp

```diff
@@ -2063,93 +2071,100 @@ void AI::KeepStation(Ship &ship, Command &command, const Body &target)
 
 void AI::Attack(Ship &ship, Command &command, const Ship &target)
 {
-	// First, figure out what your shortest-range weapon is.
-	double shortestRange = 4000.;
-	bool isArmed = false;
-	bool hasAmmo = false;
-	double minSafeDistance = 0.;
-	for(const Hardpoint &hardpoint : ship.Weapons())
-	{
-		const Weapon *weapon = hardpoint.GetOutfit();
-		if(weapon && !hardpoint.IsAntiMissile())
-		{
-			isArmed = true;
-			bool hasThisAmmo = (!weapon->Ammo() || ship.OutfitCount(weapon->Ammo()));
-			hasAmmo |= hasThisAmmo;
-
-			// Exploding weaponry that can damage this ship requires special
-			// consideration (while we have the ammo to use the weapon).
-			if(hasThisAmmo && weapon->BlastRadius() && !weapon->IsSafe())
-				minSafeDistance = max(weapon->BlastRadius() + weapon->TriggerRadius(), minSafeDistance);
-
-			// The missile boat AI should be applied at 1000 pixels range if
-			// all weapons are homing or turrets, and at 2000 if not.
-			double multiplier = (hardpoint.IsHoming() || hardpoint.IsTurret()) ? 1. : .5;
-			shortestRange = min(multiplier * weapon->Range(), shortestRange);
-		}
-	}
-	// If this ship was using the missile boat AI to run away and bombard its
-	// target from a distance, have it stop running once it is out of ammo. This
-	// is not realistic, but it's a whole lot less annoying for the player when
-	// they are trying to hunt down and kill the last missile boat in a fleet.
-	if(isArmed && !hasAmmo)
-		shortestRange = 0.;
-
+	ShipAICache &shipAICache = ship.GetAICache();
+	bool artilleryAI = shipAICache.ArtilleryAI();
+	double shortestRange = shipAICache.ShortestRange();
+	double shortestArtillery = shipAICache.ShortestArtillery();
+	double minSafeDistance = shipAICache.MinSafeDistance();
+	double totalRadius = ship.Radius() + target.Radius();
+	Point d = target.Position() - ship.Position();
+	double weaponDistance = d.Length() - totalRadius/3;
+	
 	// Deploy any fighters you are carrying.
 	if(!ship.IsYours() && ship.HasBays())
 	{
 		command |= Command::DEPLOY;
 		Deploy(ship, false);
 	}
+	
 
-	// If this ship has only long-range weapons, or some weapons have a
+	// If this ship has mostly long-range weapons, or some weapons have a
 	// blast radius, it should keep some distance instead of closing in.
-	Point d = (target.Position() + target.Velocity()) - (ship.Position() + ship.Velocity());
-	if((minSafeDistance > 0. || shortestRange > 1000.)
-			&& d.Length() < max(1.25 * minSafeDistance, .5 * shortestRange))
+	// if a weapon has blast radius, some leeway helps avoid getting hurt
+	if(minSafeDistance > 1 || (artilleryAI && shortestRange < weaponDistance))
 	{
+		minSafeDistance = 1.25 * minSafeDistance + totalRadius;
+		
+		double approachSpeed = (ship.Velocity() - target.Velocity()).Dot(d.Unit());
+		double slowdownDistance = 0;
```

Comment:
```suggestion
		double slowdownDistance = 0.;
```
It's a small thing, and it doesn't matter much but 0 is an integer, 0. is a double. So if you write it this way you will avoid it being converted every single time. Same thing with the divisions etc, I would always use 2. when dealing with doubles to avoid unnecessary casts.
You're always using doubles as far as I can see, so may as well put a `.` at the end of the numbers. (same for the `1` above, the `2` 2 below etc)